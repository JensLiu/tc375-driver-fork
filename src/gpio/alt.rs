// TODO (alepez) Remove this warning suppression
#![allow(unused)]

mod tc37x_io;
pub use tc37x_io::*;

macro_rules! extipin {
    ($( $(#[$attr:meta])* $PX:ident,)*) => {
        fn make_interrupt_source(&mut self, _syscfg: &mut $crate::syscfg::SysCfg) {
            match self {
                $(
                    $(#[$attr])*
                    Self::$PX(p) => p.make_interrupt_source(_syscfg),
                )*
                _ => {},
            }

        }

        fn trigger_on_edge(&mut self, _exti: &mut $crate::pac::EXTI, _level: $crate::gpio::Edge) {
            match self {
                $(
                    $(#[$attr])*
                    Self::$PX(p) => p.trigger_on_edge(_exti, _level),
                )*
                _ => {},
            }
        }

        fn enable_interrupt(&mut self, _exti: &mut $crate::pac::EXTI) {
            match self {
                $(
                    $(#[$attr])*
                    Self::$PX(p) => p.enable_interrupt(_exti),
                )*
                _ => {},
            }
        }
        fn disable_interrupt(&mut self, _exti: &mut $crate::pac::EXTI) {
            match self {
                $(
                    $(#[$attr])*
                    Self::$PX(p) => p.disable_interrupt(_exti),
                )*
                _ => {},
            }
        }
        fn clear_interrupt_pending_bit(&mut self) {
            match self {
                $(
                    $(#[$attr])*
                    Self::$PX(p) => p.clear_interrupt_pending_bit(),
                )*
                _ => {},
            }
        }
        fn check_interrupt(&self) -> bool {
            match self {
                $(
                    $(#[$attr])*
                    Self::$PX(p) => p.check_interrupt(),
                )*
                _ => false,
            }
        }
    };
}
use extipin;

macro_rules! pin {
    ( $($(#[$docs:meta])* <$name:ident, $Otype:ident> for $(no: $NoPin:ident,)? [$(
        $(#[$attr:meta])* $PX:ident<$A:literal $(, Speed::$Speed:ident)?>,
    )*],)*) => {
        $(
            #[derive(Debug)]
            $(#[$docs])*
            pub enum $name {
                $(
                    None($NoPin<$Otype>),
                )?

                $(
                    $(#[$attr])*
                    $PX(gpio::$PX<$crate::gpio::Alternate<$A, $Otype>>),
                )*
            }

            impl crate::Sealed for $name { }

            #[allow(unreachable_patterns)]
            impl $crate::gpio::ReadPin for $name {
                fn is_low(&self) -> bool {
                    match self {
                        $(
                            $(#[$attr])*
                            Self::$PX(p) => p.is_low(),
                        )*
                        _ => false,
                    }
                }
            }

            #[allow(unreachable_patterns)]
            impl $crate::gpio::PinSpeed for $name {
                fn set_speed(&mut self, _speed: $crate::gpio::Speed) {
                    match self {
                        $(
                            $(#[$attr])*
                            Self::$PX(p) => p.set_speed(_speed),
                        )*
                        _ => {}
                    }
                }
            }

            #[allow(unreachable_patterns)]
            impl $crate::gpio::PinPull for $name {
                fn set_internal_resistor(&mut self, _pull: $crate::gpio::Pull) {
                    match self {
                        $(
                            $(#[$attr])*
                            Self::$PX(p) => p.set_internal_resistor(_pull),
                        )*
                        _ => {}
                    }
                }
            }

            // #[allow(unreachable_patterns)]
            // impl $crate::gpio::ExtiPin for $name {
            //     extipin! { $( $(#[$attr])* $PX, )* }
            // }

            $(
                impl From<$NoPin<$Otype>> for $name {
                    fn from(p: $NoPin<$Otype>) -> Self {
                        Self::None(p)
                    }
                }
            )?

            $(
                $(#[$attr])*
                impl<MODE> From<gpio::$PX<MODE>> for $name
                where
                    MODE: $crate::gpio::marker::NotAlt + $crate::gpio::PinMode
                {
                    fn from(p: gpio::$PX<MODE>) -> Self {
                        Self::$PX(p.into_mode() $(.speed($crate::gpio::Speed::$Speed))?)
                    }
                }

                $(#[$attr])*
                impl From<gpio::$PX<$crate::gpio::Alternate<$A, $Otype>>> for $name {
                    fn from(p: gpio::$PX<$crate::gpio::Alternate<$A, $Otype>>) -> Self {
                        Self::$PX(p $(.speed($crate::gpio::Speed::$Speed))?)
                    }
                }

                $(#[$attr])*
                #[allow(irrefutable_let_patterns)]
                impl<MODE> TryFrom<$name> for gpio::$PX<MODE>
                where
                    MODE: $crate::gpio::PinMode,
                    $crate::gpio::Alternate<$A, $Otype>: $crate::gpio::PinMode,
                {
                    type Error = ();

                    fn try_from(a: $name) -> Result<Self, Self::Error> {
                        if let $name::$PX(p) = a {
                            Ok(p.into_mode())
                        } else {
                            Err(())
                        }
                    }
                }
            )*
        )*
    };

    ( $($(#[$docs:meta])* <$name:ident> default:$DefaultOtype:ident for $(no: $NoPin:ident,)? [$(
            $(#[$attr:meta])* $PX:ident<$A:literal>,
    )*],)*) => {
        $(
            #[derive(Debug)]
            $(#[$docs])*
            pub enum $name<Otype = $DefaultOtype> {
                $(
                    None($NoPin<Otype>),
                )?

                $(
                    $(#[$attr])*
                    $PX(gpio::$PX<$crate::gpio::Alternate<$A, Otype>>),
                )*
            }

            impl<Otype> crate::Sealed for $name<Otype> { }

            #[allow(unreachable_patterns)]
            impl<Otype> $crate::gpio::ReadPin for $name<Otype> {
                fn is_low(&self) -> bool {
                    match self {
                        $(
                            $(#[$attr])*
                            Self::$PX(p) => p.is_low(),
                        )*
                        _ => false,
                    }
                }
            }

            #[allow(unreachable_patterns)]
            impl<Otype> $crate::gpio::PinSpeed for $name<Otype> {
                fn set_speed(&mut self, _speed: $crate::gpio::Speed) {
                    match self {
                        $(
                            $(#[$attr])*
                            Self::$PX(p) => p.set_speed(_speed),
                        )*
                        _ => {}
                    }
                }
            }

            #[allow(unreachable_patterns)]
            impl<Otype> $crate::gpio::PinPull for $name<Otype> {
                fn set_internal_resistor(&mut self, _pull: $crate::gpio::Pull) {
                    match self {
                        $(
                            $(#[$attr])*
                            Self::$PX(p) => p.set_internal_resistor(_pull),
                        )*
                        _ => {}
                    }
                }
            }

            // #[allow(unreachable_patterns)]
            // impl<Otype> $crate::gpio::ExtiPin for $name<Otype> {
            //     extipin! { $( $(#[$attr])* $PX, )* }
            // }

            $(
                impl<Otype> From<$NoPin<Otype>> for $name<Otype> {
                    fn from(p: $NoPin<Otype>) -> Self {
                        Self::None(p)
                    }
                }
            )?

            $(
                $(#[$attr])*
                impl<MODE, Otype> From<gpio::$PX<MODE>> for $name<Otype>
                where
                    MODE: $crate::gpio::marker::NotAlt + $crate::gpio::PinMode,
                    $crate::gpio::Alternate<$A, Otype>: $crate::gpio::PinMode,
                {
                    fn from(p: gpio::$PX<MODE>) -> Self {
                        Self::$PX(p.into_mode())
                    }
                }

                $(#[$attr])*
                impl<Otype> From<gpio::$PX<$crate::gpio::Alternate<$A, Otype>>> for $name<Otype> {
                    fn from(p: gpio::$PX<$crate::gpio::Alternate<$A, Otype>>) -> Self {
                        Self::$PX(p)
                    }
                }

                $(#[$attr])*
                #[allow(irrefutable_let_patterns)]
                impl<MODE, Otype> TryFrom<$name<Otype>> for gpio::$PX<MODE>
                where
                    MODE: $crate::gpio::PinMode,
                    $crate::gpio::Alternate<$A, Otype>: $crate::gpio::PinMode,
                {
                    type Error = ();

                    fn try_from(a: $name<Otype>) -> Result<Self, Self::Error> {
                        if let $name::$PX(p) = a {
                            Ok(p.into_mode())
                        } else {
                            Err(())
                        }
                    }
                }
            )*
        )*
    };
}
use pin;

// CAN pins
pub trait CanCommon {
    type Rx;
    type Tx;
}
