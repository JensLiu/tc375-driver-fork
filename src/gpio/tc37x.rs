pub use super::Input as DefaultMode;
use super::*;

gpio!(gpio00, crate::pac::port_00::Port00, 0, P00n, [
    P00_0 : (p00_0 , 0 , [  0,     1,     2,     3,   /*4,*/   5,   /*6,*/   7  ]),
    P00_1 : (p00_1 , 1 , [  0,     1,     2,   /*3,*/   4,   /*5,*/   6,     7  ]),
    P00_2 : (p00_2 , 2 , [  0,     1,     2,   /*3,*/   4,     5,     6,     7  ]),
    P00_3 : (p00_3 , 3 , [  0,     1,     2,   /*3,*/   4,   /*5,*/   6,     7  ]),
    P00_4 : (p00_4 , 4 , [  0,     1,     2,     3,     4,   /*5,*/   6,     7  ]),
    P00_5 : (p00_5 , 5 , [  0,     1,     2,   /*3,*/ /*4,*/   5,     6,     7  ]),
    P00_6 : (p00_6 , 6 , [  0,     1,     2,     3,     4,     5,     6,     7  ]),
    P00_7 : (p00_7 , 7 , [  0,     1,     2,     3,     4,     5,     6,     7  ]),
    P00_8 : (p00_8 , 8 , [  0,     1,     2,     3,   /*4,*/   5,     6,     7  ]),
    P00_9 : (p00_9 , 9 , [  0,     1,     2,     3,     4,     5,     6,     7  ]),
    P00_10: (p00_10, 10, [  0,     1,     2,   /*3,*/ /*4,*/ /*5,*/   6,     7  ]),
    P00_11: (p00_11, 11, [  0,     1,     2,   /*3,*/   4   /*5,*/ /*6,*/ /*7*/]),
    P00_12: (p00_12, 12, [  0,     1,   /*2,*/ /*3,*/ /*4,*//*5,*/ /*6,*/    7  ]),
]);

gpio!(gpio01, crate::pac::port_01::Port01, 1, P01n, [
    P01_3 : (p01_3 , 3 , [  0,     1,     2,     3,     4,     5,     6,     7  ]),
    P01_4 : (p01_4 , 4 , [  0,     1,     2,     3,     4,     5,     6,     7  ]),
    P01_5 : (p01_5 , 5 , [  0,     1,     2,     3,     4,     5,     6,     7  ]),
    P01_6 : (p01_6 , 6 , [  0,     1,     2,     3,     4,     5,     6,     7  ]),
    P01_7 : (p01_7 , 7 , [  0,     1,     2,     3,     4,     5,     6,     7  ]),
]);


