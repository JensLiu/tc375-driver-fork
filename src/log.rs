//! Logging macros and utilities.

// Log with defmt
#[cfg(all(not(feature = "log_with_env_logger"), feature = "log_with_defmt"))]
pub use tricore_log::*;

// No log system
#[cfg(all(not(feature = "log_with_defmt"), not(feature = "log_with_env_logger")))]
pub use crate::{
    no_log_debug as debug, no_log_error as error, no_log_info as info, no_log_trace as trace,
    no_log_warn as warn,
};

// Log with log crate
#[cfg(all(not(feature = "log_with_defmt"), feature = "log_with_env_logger"))]
pub use ::log::{debug, error, info, trace, warn};

// link with defmt implementation
#[cfg(feature = "log_with_defmt")]
extern crate defmt_rtt;

#[cfg(feature = "log_with_defmt")]
mod tricore_log {
    pub use defmt::debug;
    pub use defmt::error;
    pub use defmt::info;
    pub use defmt::trace;
    pub use defmt::warn;
}

#[macro_export]
macro_rules! no_log_debug {
    ($($ignored:tt)*) => {};
}

#[macro_export]
macro_rules! no_log_trace {
    ($($ignored:tt)*) => {};
}

#[macro_export]
macro_rules! no_log_info {
    ($($ignored:tt)*) => {};
}

#[macro_export]
macro_rules! no_log_warn {
    ($($ignored:tt)*) => {};
}

#[macro_export]
macro_rules! no_log_error {
    ($($ignored:tt)*) => {};
}

/// A wrapper around a slice of bytes that implements `Display` to print the
/// bytes as a hex string.
///
/// ```
/// use bw_r_drivers_tc37x::log::HexSlice;
/// let bytes = &[1u8, 2, 3, 4];
/// assert_eq!(format!("{}", HexSlice::from(bytes)), "01020304");
/// ```
pub struct HexSlice<'a>(&'a [u8]);

impl core::fmt::Display for HexSlice<'_> {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        for byte in self.0 {
            write!(f, "{:02x}", byte)?;
        }
        Ok(())
    }
}

#[cfg(feature = "log_with_defmt")]
impl defmt::Format for HexSlice<'_> {
    fn format(&self, f: defmt::Formatter) {
        for byte in self.0 {
            defmt::write!(f, "{:02x}", byte);
        }
    }
}

impl<'a, const N: usize> From<&'a [u8; N]> for HexSlice<'a> {
    fn from(value: &'a [u8; N]) -> Self {
        HexSlice(value.as_slice())
    }
}
impl<'a> From<&'a [u8]> for HexSlice<'a> {
    fn from(value: &'a [u8]) -> Self {
        HexSlice(value)
    }
}
