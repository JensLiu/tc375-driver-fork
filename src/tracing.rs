#![allow(unused)]

extern crate std;

use std::{
    sync::{Arc, Mutex},
    vec::Vec,
};

#[cfg(feature = "defmt")]
use defmt::println;

#[cfg(not(feature = "defmt"))]
use std::println;

use crate::pac::tracing::Reporter;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ReportData {
    pub addr: usize,
    pub len: usize,
    pub action: ReportAction,
}

impl ReportData {
    pub fn new(action: ReportAction, addr: usize, len: usize) -> Self {
        Self { addr, len, action }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ReportAction {
    Read,
    Write(u32),
    LoadModifyStore(u64),
}

// TODO Remove, should be injectable
pub const READ_VALUE: u32 = 0x0;

#[derive(Default, Clone)]
pub struct LogEffectReporter(Arc<Mutex<Vec<ReportData>>>);

impl LogEffectReporter {
    pub fn get_logs(&self) -> Vec<ReportData> {
        let mut g = self.0.lock().unwrap();
        let len = g.len();
        g.drain(0..len).collect()
    }

    fn push(&self, report: ReportData) {
        self.0.lock().unwrap().push(report);
    }

    fn report(&self, action: ReportAction, addr: usize, len: usize) {
        self.push(ReportData::new(action, addr, len))
    }
}

impl Reporter for LogEffectReporter {
    fn read_volatile(&self, addr: usize, len: usize) -> u32 {
        self.report(ReportAction::Read, addr, len);
        READ_VALUE
    }

    fn write_volatile(&self, addr: usize, len: usize, val: u32) {
        self.report(ReportAction::Write(val), addr, len);
    }

    fn load_modify_store(&self, addr: usize, val: u64) {
        self.report(
            ReportAction::LoadModifyStore(val),
            addr,
            core::mem::size_of::<u64>(),
        );
    }
}

#[derive(Default, Copy, Clone)]
pub struct PrintEffectReporter;

impl Reporter for PrintEffectReporter {
    fn read_volatile(&self, ptr: usize, len: usize) -> u32 {
        println!("read_volatile 0x{:08X} len={}", ptr, len);
        READ_VALUE
    }

    fn write_volatile(&self, ptr: usize, len: usize, val: u32) {
        println!("write_volatile 0x{:08X} len={} val={}", ptr, len, val);
    }

    fn load_modify_store(&self, ptr: usize, val: u64) {
        println!("load_modify_store 0x{:08X} val={}", ptr, val);
    }
}

pub fn redirect_to_print() {
    use std::boxed::Box;
    use crate::pac;
    let reporter = PrintEffectReporter::default();
    pac::tracing::set_effect_reporter(Box::new(reporter.clone()));
}
