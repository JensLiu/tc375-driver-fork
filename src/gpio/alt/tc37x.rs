use super::*;
use crate::gpio::{self, NoPin, OpenDrain, PushPull};

pub mod can0 {
    use super::*;

    pin! {
        <Rx, PushPull> for no:NoPin, [
            // FIXME Alternate function here?
            P20_7<1>, // CAN00:RXDB = P20.7:IN
        ],

        <Tx, PushPull> for no:NoPin, [
            P20_8<5>, // CAN00:TXD = P20.8:ALT(5)
        ],
    }

    impl CanCommon for crate::pac::can0::Can0 {
        type Rx = Rx;
        type Tx = Tx;
    }
}
