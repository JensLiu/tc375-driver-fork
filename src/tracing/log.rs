#![allow(clippy::expect_used)]

use super::TraceGuard;
use crate::tracing::{LoadModifyStoreEntry, ReadEntry, ReportEntry, WriteEntry};
use std::any::Any;
use std::collections::{HashMap, VecDeque};
use std::fmt::{Debug, Display, Formatter, Write};
use std::sync::{Arc, Mutex, MutexGuard};

struct ReadFifoEntry {
    addr: usize,
    len: usize,
    val: u64,
}

#[derive(Default)]
struct ReadFifo(VecDeque<ReadFifoEntry>);

#[derive(Default)]
struct SharedData {
    log: Log,
    read_fifo: ReadFifo,
}

struct Reporter {
    shared_data: Arc<Mutex<SharedData>>,
}

pub struct Report {
    shared_data: Arc<Mutex<SharedData>>,
    _guard: TraceGuard,
}

impl Default for Report {
    fn default() -> Self {
        Self::new()
    }
}

impl Report {
    pub fn new() -> Self {
        let data = Arc::new(Mutex::new(SharedData::default()));
        let reporter = Reporter {
            shared_data: Arc::clone(&data),
        };
        let guard = TraceGuard::new(reporter);
        Self {
            shared_data: data,
            _guard: guard,
        }
    }

    pub fn take_log(&self) -> Log {
        let mut g = self.shared_data();
        let len = g.log.0.len();
        Log(g.log.0.drain(0..len).collect())
    }

    pub fn expect_read(&self, addr: usize, len: usize, val: u64) {
        self.shared_data()
            .read_fifo
            .0
            .push_front(ReadFifoEntry { addr, len, val })
    }

    pub fn comment(&self, s: impl Into<String>) {
        self.shared_data().log.0.push(LogEntry::Comment(s.into()));
    }

    fn shared_data(&self) -> MutexGuard<SharedData> {
        self.shared_data.lock().unwrap()
    }
}

impl Reporter {
    fn push(&self, report: ReportEntry) {
        self.shared_data().log.0.push(LogEntry::ReportEntry(report));
    }

    fn shared_data(&self) -> MutexGuard<SharedData> {
        self.shared_data.lock().unwrap()
    }
}

impl super::Reporter for Reporter {
    fn read_volatile(&self, addr: usize, len: usize) -> u64 {
        let entry = self
            .shared_data()
            .read_fifo
            .0
            .pop_front()
            .expect("Unexpected read");

        if entry.addr == addr && entry.len == len {
            let val = entry.val;
            self.push(ReportEntry::Read(ReadEntry { addr, len, val }));
            val
        } else {
            panic!("Unexpected read at address 0x{:08X} and len {}", addr, len)
        }
    }

    fn write_volatile(&self, addr: usize, len: usize, val: u64) {
        self.push(ReportEntry::Write(WriteEntry { addr, len, val }));
    }

    fn load_modify_store(&self, addr: usize, val: u64) {
        self.push(ReportEntry::LoadModifyStore(LoadModifyStoreEntry {
            addr,
            val,
        }));
    }
}

impl Drop for Reporter {
    fn drop(&mut self) {
        if !self.shared_data().read_fifo.0.is_empty() {
            panic!("More read where expected");
        }
    }
}

#[derive(Default, Debug, PartialEq)]
pub struct Log(Vec<LogEntry>);

#[derive(Debug, PartialEq)]
enum LogEntry {
    Comment(String),
    ReportEntry(ReportEntry),
}

impl Display for Log {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        for entry in &self.0 {
            match entry {
                LogEntry::ReportEntry(ReportEntry::Read(x)) => {
                    write!(f, "r    0x{:08X} {:02} 0x{:08X}", x.addr, x.len, x.val);
                }
                LogEntry::ReportEntry(ReportEntry::Write(x)) => {
                    write!(f, "w    0x{:08X} {:02} 0x{:08X}", x.addr, x.len, x.val);
                }
                LogEntry::ReportEntry(ReportEntry::LoadModifyStore(x)) => {
                    let mask = (x.val >> 32);
                    let val = (x.val & 0xFFFFFFFF);
                    write!(f, "ldms 0x{:08X} 0x{:08X} 0x{:08X}", x.addr, mask, val);
                }
                LogEntry::Comment(s) => {
                    write!(f, "# {s}");
                }
            }

            f.write_char('\n');
        }

        Ok(())
    }
}
