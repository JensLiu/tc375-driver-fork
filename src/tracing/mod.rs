#![allow(unused)]

pub mod log;
pub mod print;

extern crate std;

use std::{
    sync::{Arc, Mutex},
    vec::Vec,
};

use crate::pac::tracing::Reporter;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ReportData {
    pub addr: usize,
    pub len: usize,
    pub action: ReportAction,
}

impl ReportData {
    pub fn new(action: ReportAction, addr: usize, len: usize) -> Self {
        Self { addr, len, action }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ReportAction {
    Read,
    Write(u32),
    LoadModifyStore(u64),
}
