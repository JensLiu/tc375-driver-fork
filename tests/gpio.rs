use embedded_hal::digital::PinState;
use tc37x::P01;
use tc37x::{P00, P20};

use bw_r_driver_tc37x::gpio::{ErasedPin, GpioExt};
use bw_r_driver_tc37x::tracing;
use tracing::log::Report;

#[test]
fn test_pin_set_high_and_low() {
    let report = Report::new();

    let port = P00.split();
    let mut output = port.p00_5.into_push_pull_output();

    output.set_high();
    output.set_low();

    insta::assert_snapshot!(report.take_log());
}

#[test]
fn test_pin_set_two_pins_same_port_high() {
    let report = Report::new();

    let port = P00.split();
    let mut p00_5 = port.p00_5.into_push_pull_output();
    let mut p00_6 = port.p00_6.into_push_pull_output();

    p00_5.set_high();
    p00_6.set_high();

    insta::assert_snapshot!(report.take_log());
}

#[test]
fn test_pin_set_two_pins_on_two_ports_high() {
    let report = Report::new();

    let port = P00.split();
    let gpio01 = P01.split();
    let mut p00_5 = port.p00_5.into_push_pull_output();
    let mut p01_7 = gpio01.p01_7.into_push_pull_output();

    p00_5.set_high();
    p01_7.set_high();

    insta::assert_snapshot!(report.take_log());
}

#[test]
fn avoid_side_effects_when_mode_does_not_change() {
    let report = Report::new();

    let port = P00.split();
    let pin = port.p00_5.into_push_pull_output();
    let _pin = pin.into_push_pull_output();

    insta::assert_snapshot!(report.take_log());
}

#[test]
fn test_input_pin() {
    let report = Report::new();

    let port = P00.split();
    let pin = port.p00_7.into_input();

    report.expect_read(0xF003A024, 4, 0b00000000000000000000000000000000);
    let is_high = pin.is_high();
    assert!(!is_high);

    report.expect_read(0xF003A024, 4, 0b00000000000000000000000010000000);
    let is_high = pin.is_high();
    assert!(is_high);

    insta::assert_snapshot!(report.take_log());
}

#[test]
fn test_input_pin_pull_up() {
    let _report = Report::new();

    let port = P00.split();
    let _pin = port.p00_7.into_pull_up_input();

    // TODO Review report. Not sure about PCx
    // insta::assert_snapshot!(report.take_log());
}

#[test]
fn test_output_pin_type_erasure_number() {
    let report = Report::new();

    let port = P00.split();
    let output = port.p00_5.into_push_pull_output();
    let mut output = output.erase_number();

    output.set_high();
    output.set_low();

    insta::assert_snapshot!(report.take_log());
}

#[test]
fn test_output_pin_type_erasure_port_and_number() {
    let report = Report::new();

    let port = P00.split();
    let output = port.p00_5.into_push_pull_output();
    let mut output = output.erase();

    output.set_high();
    output.set_low();

    insta::assert_snapshot!(report.take_log());
}

#[test]
fn test_input_pin_type_erasure_number() {
    let report = Report::new();

    let port = P00.split();
    let pin = port.p00_7.into_input();
    let pin = pin.erase_number();

    report.expect_read(0xF003A024, 4, 0b00000000000000000000000000000000);
    let is_high = pin.is_high();
    assert!(!is_high);

    report.expect_read(0xF003A024, 4, 0b00000000000000000000000010000000);
    let is_high = pin.is_high();
    assert!(is_high);

    insta::assert_snapshot!(report.take_log());
}

#[test]
fn test_input_pin_type_erasure_port_and_number() {
    let report = Report::new();

    let port = P00.split();
    let pin = port.p00_7.into_input();
    let pin = pin.erase();

    report.expect_read(0xF003A024, 4, 0b00000000000000000000000000000000);
    let is_high = pin.is_high();
    assert!(!is_high);

    report.expect_read(0xF003A024, 4, 0b00000000000000000000000010000000);
    let is_high = pin.is_high();
    assert!(is_high);

    insta::assert_snapshot!(report.take_log());
}

#[test]
fn toggle_output_pin() {
    let report = Report::new();

    let port = P00.split();
    let mut output = port.p00_5.into_push_pull_output();

    output.toggle();

    insta::assert_snapshot!(report.take_log());
}

#[test]
fn toggle_output_pin_type_erasure_number() {
    let report = Report::new();

    let port = P00.split();
    let output = port.p00_5.into_push_pull_output();
    let mut output = output.erase_number();

    output.toggle();

    insta::assert_snapshot!(report.take_log());
}

#[test]
fn toggle_output_pin_type_erasure_port_and_number() {
    let report = Report::new();

    let port = P00.split();
    let output = port.p00_5.into_push_pull_output();
    let mut output = output.erase();

    output.toggle();

    insta::assert_snapshot!(report.take_log());
}

// Set the pin high by using StatefulOutputPin interface
// This is needed to ensure the StatefulOutputPin trait is implemented
fn stateful_output_pin_is_set_high(pin: impl embedded_hal::digital::StatefulOutputPin) -> bool {
    pin.is_set_high().unwrap()
}

#[test]
fn toggle_stateful_output_pin_stateful() {
    let report = Report::new();

    let port = P00.split();
    let output = port.p00_5.into_push_pull_output();

    report.expect_read(0xF003A000, 4, 0b00000000000000000000000000100000);

    assert!(stateful_output_pin_is_set_high(output));

    insta::assert_snapshot!(report.take_log());
}

#[test]
fn toggle_stateful_output_pin_type_erasure_number() {
    let report = Report::new();

    let port = P00.split();
    let output = port.p00_5.into_push_pull_output();
    let output = output.erase_number();

    report.expect_read(0xF003A000, 4, 0b00000000000000000000000000100000);

    assert!(stateful_output_pin_is_set_high(output));

    insta::assert_snapshot!(report.take_log());
}

#[test]
fn toggle_stateful_output_pin_type_erasure_port_and_number() {
    let report = Report::new();

    let port = P00.split();
    let output = port.p00_5.into_push_pull_output();
    let output = output.erase();

    report.expect_read(0xF003A000, 4, 0b00000000000000000000000000100000);

    assert!(stateful_output_pin_is_set_high(output));

    insta::assert_snapshot!(report.take_log());
}

#[test]
fn type_erasure_with_into() {
    let _report = Report::new();

    let port = P00.split();
    let output = port.p00_5.into_push_pull_output();
    let output = output.erase();
    let _output: ErasedPin<_> = output;
}

#[test]
fn pin_can_type_match_with_peripheral() {
    let _report = Report::new();

    use self::mock_can::*;

    let port = P20.split();
    let rx = port.p20_7;
    let tx = port.p20_8;

    let _can = Can::<tc37x::can0::Can0>::new(tx, rx);
}

mod mock_can {
    use bw_r_driver_tc37x::gpio::*;

    pub struct Can<CAN: alt::CanCommon> {
        _tx_pin: CAN::Tx,
        _rx_pin: CAN::Rx,
    }

    impl<CAN: alt::CanCommon> Can<CAN> {
        pub fn new(tx_pin: impl Into<CAN::Tx>, rx_pin: impl Into<CAN::Rx>) -> Self {
            Can {
                _tx_pin: tx_pin.into(),
                _rx_pin: rx_pin.into(),
            }
        }
    }
}

// TODO Discussed during meeting 2023-11-24
// User case is: I want to control many pins in the same port at once, e.g. to
// implement bit banging.
// #[test]
// fn test_gpio_syncronous_update_all_pins_same_mode() {
//     let report = Report::new();
//
//     let port = P00.split();
//     let pins = (port.p00_1, port.p00_6, port.p00_7);
//     let group = PinGroup::new(pins);
//     let group = group.into_push_pull_output();
//     group.set_high();
//     group.set_low();
//     group.set_raw(0b1100010);
//     group.set_values((true, true, true));
// }
//
// #[test]
// fn test_gpio_syncronous_update_mixed_pins_mode() {
//     let report = Report::new();
//
//     let port = P00.split();
//
//     let pins = (
//         port.p00_1.into_input(),
//         port.p00_6.into_push_pull_output(),
//         port.p00_7.into_push_pull_output(),
//     );
//
//     let group = PinGroup::new(pins);
//
//     // TODO Check if setting an input pin is not UB
//     // Only 6 and 7 should be set
//     group.set_high();
//     group.set_low();
//     group.set_raw(0b1100010);
//     group.set_values((true, true, true));
// }

// User case is: I want to control many pins in the same port at once, e.g. to
// implement bit banging.
#[test]
fn test_gpio_outport_array() {
    use bw_r_driver_tc37x::gpio::group::PinArray;

    let report = Report::new();
    let port = P00.split();

    let mut group = PinArray([
        port.p00_1.into_push_pull_output().erase_number(),
        port.p00_6.into_push_pull_output().erase_number(),
        port.p00_7.into_push_pull_output().erase_number(),
    ]);

    group.set_high();
    group.set_state([PinState::High, PinState::High, PinState::High]);
    group.set_low();
    group.set_state([PinState::Low, PinState::Low, PinState::Low]);

    insta::assert_snapshot!(report.take_log());
}

// User case is: I want to control many pins in the same port at once, e.g. to
// implement bit banging.
#[test]
fn test_gpio_outport_tuple() {
    use bw_r_driver_tc37x::gpio::group::PinGroup;

    let report = Report::new();

    report.comment("Configure pins");

    let port = P00.split();

    let mut group = (
        port.p00_1.into_push_pull_output(),
        port.p00_6.into_push_pull_output(),
        port.p00_7.into_push_pull_output(),
    )
        .into_pin_group();

    report.comment("Set all pins high");
    group.set_high();

    report.comment("Set all pins low");
    group.set_low();

    insta::assert_snapshot!(report.take_log());

    // Make sure set_high is equivalent to set_state and write with all pins high
    {
        group.set_high();
        let set_high_log = report.take_log();

        group.set_state([PinState::High, PinState::High, PinState::High]);
        let set_state_log = report.take_log();

        group.write(0xFFFFFFFF);
        let write_log = report.take_log();

        assert_eq!(set_high_log, set_state_log);
        assert_eq!(set_high_log, write_log);
    }

    // Make sure set_low is equivalent to set_state and write with all pins low
    {
        group.set_low();
        let set_low_log = report.take_log();

        group.set_state([PinState::Low, PinState::Low, PinState::Low]);
        let set_state_log = report.take_log();

        group.write(0x00000000);
        let write_log = report.take_log();

        assert_eq!(set_low_log, set_state_log);
        assert_eq!(set_low_log, write_log);
    }
}
