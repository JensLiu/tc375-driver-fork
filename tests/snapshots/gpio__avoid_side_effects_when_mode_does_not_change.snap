---
source: tests/gpio.rs
expression: report.take_log()
---
w    0xF003A004 04 0x00200000
ldms 0xF003A014 0x0000F800 0x00008000
w    0xF003A004 04 0x00200000

